### ISSUE TITLE
YAML parser rejects Unicode surrogates.

### ISSUE THREAD

**Issue by lhmathies** [mikefarah/yq#2206]
**Describe the bug**

The YAML parser rejects Unicode surrogates, but the JSON parser accepts them. This breaks the expectation that you can parse JSON as if it were YAML.

I have a real world example where the JSON response body from an HTTP call to an API endpoint contains surrogates (in string values), but I'll illustrate it with a single non-BMP character (u+10336 Gothic Letter Iuja, 𐌶):

$ yq --version
yq (https://github.com/mikefarah/yq/) version v4.44.5
$ echo -n '"\ud800\udf36"' | yq -py
Error: bad file '-': yaml: found invalid Unicode character escape code
$ echo -n '"\ud800\udf36"' | yq -pj
𐌶
$ echo -n '"\ud800\udf36"' | yq -pj | tr -d \\n | iconv -t utf-32le | od -t x4
0000000 00010336
0000004

Version of yq: 4.44.5
Operating system: linux amd64
Installed via: binary release

**Comment by lhmathies** [comment]
YAML 1,2 specification, first line of section 5.2:

``` text
 All characters mentioned in this specification are Unicode code points. Each such code point is written as one or more bytes depending on the character encoding used. Note that in UTF-16, characters above #xFFFF are written as four bytes, using a surrogate pair.
```

**Comment by lhmathies** [comment]
To be more explicit: The only spec-compliant way to represent 𐌶 in JSON is as `"\ud800\udf36"`, unless your input is UTF-8. So why the API I have to deal with doesn't just do that is a good question. (The contest is a user upload; all non-ASCII chars are converted to `\uxxxx` escapes, but the JSON is served with an explicit `content-type: text/plain; charset=utf-8`(!) My guess is that the server stores some type of UTF-16 and whoever coded the read endpoint was unclear on the concepts; handing the same string to JS code \[which is the site's main way of presenting the content\] does after all work).

So regardless of what the YAML spec says or doesn't say about surrogates in `\udxxx` representation, it does say that it's intended to be a superset of JSON; it's surprising that `yq -py` doesn't accept it.

**Comment by mikefarah** [comment]
Yeah this is a known issue with the underlying go-yaml library, which hasn't received much love recently :(
